Refactor `-[WKContentViewInteraction _interpretKeyEvent:]` to additionally receive the relevant scrolling node ID

rr-codes opened the pull request:
#### 75aceefef09dbf75d31fb0814930e11349f2f81f
<pre>
Refactor `-[WKContentViewInteraction _interpretKeyEvent:]` to additionally receive the relevant scrolling node ID
<a href="https://bugs.webkit.org/show_bug.cgi?id=287723">https://bugs.webkit.org/show_bug.cgi?id=287723</a>
<a href="https://rdar.apple.com/144880163">rdar://144880163</a>

Reviewed by Aditya Keerthi and Abrar Rahman Protyasha.

To support scrolling sub-scrollable regions using keyboard scrolling on iOS, the `_interpretKeyEvent` method
on `WKContentViewInteraction` needs to know which scroll view to scroll, and so the scrolling node ID need to be
provided to this method.

To faciliate this, a new `KeyEventInterpretationContext` type is introduced that contains the relevant information
needed to interpret the key event. Currently, this is the existing `isCharEvent` property, in addition to the new
`scrollingNode` property.

This also involves refactoring and exposing a new `focusedScrollableArea` function on `EventHandler` which is used to get
the relevant scrolling node ID.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::enclosingScrollableArea const):
(WebCore::EventHandler::focusedScrollableArea const):
(WebCore::EventHandler::shouldUseSmoothKeyboardScrollingForFocusedScrollableArea):
(WebCore::EventHandler::enclosingScrollableArea): Deleted.
* Source/WebCore/page/EventHandler.h:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/KeyEventInterpretationContext.cpp: Added.
(WebKit::operator&lt;&lt;):
* Source/WebKit/Shared/KeyEventInterpretationContext.h: Added.
* Source/WebKit/Shared/KeyEventInterpretationContext.serialization.in: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::interpretKeyEvent):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _interpretKeyEvent:withContext:]):
(-[WKContentView _interpretKeyEvent:isCharEvent:]): Deleted.
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::interpretKeyEvent):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::handleEditingKeyboardEvent):
* Tools/SwiftBrowser/SwiftBrowser.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/290428@main">https://commits.webkit.org/290428@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d58534d17fe56eded35dab7851c5f8af9a1f64a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89998 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9527 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44898 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94998 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/40771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92050 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9914 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17813 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/69286 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/40771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92999 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/7588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81643 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49643 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/7316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/36009 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39905 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/77651 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37069 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96824 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17186 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17442 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77466 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19134 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21936 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/10380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17196 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22521 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16937 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20389 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/18720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Tools/SwiftBrowser/SwiftBrowser.xcodeproj/project.pbxproj:
```diff
@@ -300,13 +300,19 @@
 			buildSettings = {
+				SUPPORTED_PLATFORMS = "iphoneos iphonesimulator macosx xros xrsimulator";
+				SUPPORTS_MACCATALYST = NO;
+				TARGETED_DEVICE_FAMILY = "1,2";
```
Ideally, a separate commit. But I guess it's fine...

aprotyas commented on Source/WebKit/Shared/KeyEventInterpretationContext.h:
```diff
@@ -0,0 +1,46 @@
+
+namespace WTF {
+class TextStream;
+};
```
You don't need the semicolon here
```suggestion
}
```

aprotyas commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -615,8 +616,23 @@ static inline FloatRect adjustExposedRectForNewScale(const FloatRect& exposedRec
 
     updateLastNodeBeforeWritingSuggestions(event);
 
+    auto scrollingNodeID = [&]() -> std::optional<WebCore::ScrollingNodeID> {
```
Nit: you can drop the empty arg list
```suggestion
    auto scrollingNodeID = [&] -> std::optional<WebCore::ScrollingNodeID> {
```

aprotyas commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -615,8 +616,23 @@ static inline FloatRect adjustExposedRectForNewScale(const FloatRect& exposedRec
+        if (!frame)
+            return std::nullopt;
+
+        auto scrollableArea = frame->eventHandler().focusedScrollableArea();
```
Explicit smart ptr

rr-codes committed:
Refactor `-[WKContentViewInteraction _interpretKeyEvent:]` to additionally receive the relevant scrolling node ID
https://bugs.webkit.org/show_bug.cgi?id=287723
rdar://144880163

Reviewed by Aditya Keerthi and Abrar Rahman Protyasha.

To support scrolling sub-scrollable regions using keyboard scrolling on iOS, the `_interpretKeyEvent` method
on `WKContentViewInteraction` needs to know which scroll view to scroll, and so the scrolling node ID need to be
provided to this method.

To faciliate this, a new `KeyEventInterpretationContext` type is introduced that contains the relevant information
needed to interpret the key event. Currently, this is the existing `isCharEvent` property, in addition to the new
`scrollingNode` property.

This also involves refactoring and exposing a new `focusedScrollableArea` function on `EventHandler` which is used to get
the relevant scrolling node ID.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::enclosingScrollableArea const):
(WebCore::EventHandler::focusedScrollableArea const):
(WebCore::EventHandler::shouldUseSmoothKeyboardScrollingForFocusedScrollableArea):
(WebCore::EventHandler::enclosingScrollableArea): Deleted.
* Source/WebCore/page/EventHandler.h:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/KeyEventInterpretationContext.cpp: Added.
(WebKit::operator<<):
* Source/WebKit/Shared/KeyEventInterpretationContext.h: Added.
* Source/WebKit/Shared/KeyEventInterpretationContext.serialization.in: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::interpretKeyEvent):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _interpretKeyEvent:withContext:]):
(-[WKContentView _interpretKeyEvent:isCharEvent:]): Deleted.
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::interpretKeyEvent):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::handleEditingKeyboardEvent):
* Tools/SwiftBrowser/SwiftBrowser.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/290428@main
